src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -291,50 +326,79 @@ void registerNext() {
       }
     }
 
+    private void clearKnobsAfter(int i) {
+      while (i < this.knobs.length) {
+        sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
+        ++i;
+      }
+    }
+
     void register(LXDeviceComponent device) {
-      if (this.device != device) {
-        unregister();
-        this.device = device;
-        if (this.device instanceof LXEffect) {
-          this.effect = (LXEffect) this.device;
-          this.effect.enabled.addListener(this);
-        } else if (this.device instanceof LXPattern) {
-          this.pattern = (LXPattern) this.device;
+      if (this.device == device) {
+        return;
+      }
+      unregister();
+      this.device = device;
+
+      if (this.device == null) {
+        clearKnobsAfter(0);
+        return;
+      } else if (this.device instanceof LXEffect) {
+        this.effect = (LXEffect) this.device;
+        this.effect.enabled.addListener(this);
+      } else if (this.device instanceof LXPattern) {
+        this.pattern = (LXPattern) this.device;
+      }
+
+      this.device.controlSurfaceSemaphore.increment();
+
+      boolean isEnabled = false;
+      if (this.effect != null) {
+        isEnabled = this.effect.isEnabled();
+      } else if (this.pattern != null) {
+        isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
+      }
+      sendNoteOn(0, DEVICE_ON_OFF, isEnabled ? LED_ON : LED_OFF);
+
+      int i = 0;
+      Set<AggregateParameter> knownParents = new HashSet<>();
+      for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
+        if (parameter == null) {
```

Comment:
Feels like this can still throw a `NullPointerException` and crash if there is an explicit `null` in the list of knobs while `i >= this.knobs.length`